ui/src/criteria/attribute.tsx

```diff
@@ -81,23 +133,145 @@ function AttributeEdit(props: AttributeEditProps) {
   const hintDisplayName = (hint: tanagra.EnumHintValue) =>
     hint.displayName || "Unknown Value";
 
-  const enumHintValues = underlay.entities
+  const attributeFilterHint = underlay.entities
     .find((g) => g.name === underlay.primaryEntity)
-    ?.attributes?.find((attribute) => attribute.name === props.data.attribute)
-    ?.attributeFilterHint?.enumHint?.enumHintValues;
+    ?.attributes?.find(
+      (attribute) => attribute.name === props.data.attribute
+    )?.attributeFilterHint;
+  const enumHintValues = attributeFilterHint?.enumHint?.enumHintValues;
+  const integerBoundsHint = attributeFilterHint?.integerBoundsHint;
 
-  const selectionIndex = (hint: tanagra.EnumHintValue) =>
-    props.data.selected.findIndex(
-      (row) => row.id === hint.attributeValue?.int64Val
-    );
+  const hasValidIntegerBounds =
+    isValid(integerBoundsHint?.min) && isValid(integerBoundsHint?.max);
 
-  if (enumHintValues?.length === 0) {
+  if (!hasValidIntegerBounds && enumHintValues?.length === 0) {
     return (
       <Typography>
         No information for attribute {props.data.attribute}.
       </Typography>
     );
   }
+
+  if (hasValidIntegerBounds) {
+    const minBound = integerBoundsHint?.min || 0; // This is to ensure the compiler won't complain the object might be undefined
+    const maxBound = integerBoundsHint?.max || 0; // although we already know that min and max is valid
+
+    //two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number>(minBound);
+    const [maxInputValue, setMaxInputValue] = useState<number>(maxBound);
+    const [minValue, setMinValue] = useState<number>(minBound);
+    const [maxValue, setMaxValue] = useState<number>(maxBound);
+
+    const updateValue = (newMin: number, newMax: number) => {
+      setMinValue(newMin);
+      setMaxValue(newMax);
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
```

Comment:
This function has a third parameter `activeThumb` which I think will make this code easier to understand and allow better handling of the edge cases. Look at the code for this example: https://mui.com/material-ui/react-slider/#minimum-distance

Reply:
Hi Tim, I look at the `activeThumb` parameter. It actually tells the `handleChange` callback which slider knob the user is dragging. It will be useful if we are going to shift the range when the slider knob reach the minimum distance. However, for our case, we don't need to shift the range, but we need to make sure the slider knob won't cross each other, i.e., there is a `minimumDistance  = 0`. I already set `disableSwap` to achieve this. So I think `activeThumb` won't be useful in our case. The edge cases we are handling are from the input box actually, which is not related to the slider and `activeThumb` either. We can have a quick chat on this also.

Reply:
I thought knowing which thumb was being dragged might be helpful but if it's not that's okay.